### PR TITLE
NRQL Expected Values: Add bounded comparison and tag interpolation

### PR DESCRIPTION
**Pull request by dbudziwojskiNR**
This PR extends the functionality of NRQL Expected value tests. Specifically, this PR add the ability to assert that a value is greater, lesser, or between configured expected values. It also adds the ability to use the test tag within a NRQL query.

### CODE REVIEW

**Review by sigilioso — COMMENTED** [review]
Thanks for the contribution! I've left some comments.

File: internal/newrelic/client.go
Comment (dbudziwojskiNR) [reply internal/newrelic/client.go:161]: Fixed :)

File: internal/spec/definition.go
Comment (dbudziwojskiNR) [reply internal/spec/definition.go:53]: I've updated these to a *float64 to allow for a nil float if the user does not want to provide bounds.

File: internal/spec/definition.go
Comment (dbudziwojskiNR) [reply internal/spec/definition.go:51]: Using the `any` type allows us to compare the expected value and actual values using native type comparisons (i.e. `3==5`) rather than converting them both to string and doing a string comparison (`"3"=="5"`).

File: internal/newrelic/client_test.go
Comment (sigilioso) [inline internal/newrelic/client_test.go:261]: Could we also include test cases for different types? For instance, should `actualResult: "10.0"` work?

**Review by sigilioso — COMMENTED** [review]
Good job so far and thanks for addressing the comments.

File: internal/spec/definition.go
Comment (sigilioso) [reply internal/spec/definition.go]: I see. Then, can we document this behavior somewhere? I think it is not obvious when you are working on the test definition. For instance, according to [this function](https://github.com/newrelic/newrelic-integration-e2e-action/pull/34/files#diff-79a0456242c011e44009a8d20c0446a7dd16aef7ff7382b972883822e8cfed79R192), booleans will work no matter if they are bools or strings but it is not the same for numbers, is it?

File: internal/newrelic/client.go
Comment (csongnr) [inline internal/newrelic/client.go:175]: should add a short circuit error that checks either before line 175 here, or first thing in the function `checkBounds` to make sure that at least either `LowerBoundedValue` or `UpperBoundedValue` are populated.

Comment (dbudziwojskiNR) [reply internal/newrelic/client.go:175]: We actually do ensure that one or both are populated. So, this function is never called if that is the case.

File: internal/newrelic/client.go
Comment (csongnr) [reply internal/newrelic/client.go:212]: ^ i think this is a good suggestion, but might be a little easier to understand if written this way: 
```
// if either lowerBound is nil and result is less than upper bound, OR upperbound is nil and result is greater than lower bound, return nil, otherwise return an error
if (lowerBound == nil && *upperBound >= result) || (upperBound == nil && result <= *lowerBound) {
       return nil
}
rangeAsString := formatRange(lowerBound, upperBound) // returns "[-Inf, x]" "[x, Inf]" or "[x, y]" depending of bounds
return fmt.Errorf("%w - expected value in range %s", ErrAssertionFailure, rangeAsString)
```

File: internal/newrelic/client.go
Comment (csongnr) [reply internal/newrelic/client.go:175]: are you referring to the check that is happening at the end of the `checkBounds` function?

Comment (csongnr) [reply internal/newrelic/client.go:175]: oh wait, you said this function is never called if thats the case, maybe i haven't gotten to that part of the pr yet

File: internal/newrelic/client.go
Comment (csongnr) [inline internal/newrelic/client.go:239]: ah i see where it happens now. OK, please remove this error as well

File: internal/newrelic/client.go
Comment (csongnr) [reply internal/newrelic/client.go:212]: i agree w/ the suggestion above, w/ the caveat that i think it would be a little bit more straightforward if written this way, because its easier to add a comment on what the expected behavior is, (and also i think the conditional for checking lowerBound against the result was swapped, should be `<=` rather than `>=`) but its a personal preference: 
```
// if either lowerBound is nil or lowerBound <= result, AND either upperBound is nil or result <= upperBound, return nil, otherwise, error 
if (lowerBound == nil || *lowerBound <= result) && (upperBound == nil || result <= *upperBound) {
       return nil
}
rangeAsString := formatRange(lowerBound, upperBound) // returns "[-Inf, x]" "[x, Inf]" or "[x, y]" depending of bounds
return fmt.Errorf("%w - expected value in range %s", ErrAssertionFailure, rangeAsString)
```

Comment (dbudziwojskiNR) [reply internal/newrelic/client.go]: Looks good to me. Updated the PR.

File: internal/newrelic/client_test.go
Comment (dbudziwojskiNR) [reply internal/newrelic/client_test.go]: This would not work since bounded comparisons only support ints/floats.

File: internal/spec/definition.go
Comment (dbudziwojskiNR) [reply internal/spec/definition.go]: Fair point. The reason we currently treat false and "false" as the same is because the NRQL API seems to not return a native boolean, rather it returns a string boolean. How would you like to document this?

File: internal/newrelic/client.go
Comment (dbudziwojskiNR) [reply internal/newrelic/client.go]: I did some refactoring into a separate file. Have a look and let me know your thoughts :)

File: internal/spec/definition.go
Comment (sigilioso) [reply internal/spec/definition.go]: Some notes could be added to the [README's spec definition](https://github.com/newrelic/newrelic-integration-e2e-action#spec-file-for-the-e2e). Moreover, a comment in the function handling the value (and its type) can be useful too.

File: internal/newrelic/asserter.go
Comment (sigilioso) [inline internal/newrelic/asserter.go:119]: Can we remove this comment? It uses almost the same syntax as the actual code below so I don't think it helps to clarify.

File: internal/newrelic/asserter.go
Comment (sigilioso) [inline internal/newrelic/asserter.go:127]:
```suggestion
// formatRange returns "[-Inf, x]" "[x, Inf]" or "[x, y]" depending of bounds
```
https://tip.golang.org/doc/comment

File: internal/spec/definition.go
Comment (sigilioso) [inline internal/spec/definition.go:106]: Thanks for moving the definition validation here. Out of curiosity, why did you choose to iterate the scenarios and nrqls instead of leveraging it to the YML unmarshaler by [customizing the unmarshal behavior](https://pkg.go.dev/gopkg.in/yaml.v3?utm_source=godoc#Unmarshaler)?

**Review by sigilioso — COMMENTED** [review]
Great progress! I've left a few new comments, mostly _nits_.

File: internal/newrelic/asserter.go
Comment (dbudziwojskiNR) [reply internal/newrelic/asserter.go:1]: Went ahead and moved it back and correctly handled any errors. Agreed it looks better :slightly_smiling_face: Regarding your second point, I'm not sure what you mean by a specific struct? If that's referring to creating a specific error struct, I feel like that might be a bit overkill for most use cases, as an error (of error type) feels more than sufficient.

**Review by csongnr — APPROVED** [review]
LGTM :) good job daniel thanks for all your hard work. please wait for approval from core-int before merging, thanks

File: internal/newrelic/asserter.go
Comment (sigilioso) [reply internal/newrelic/asserter.go:1]: Thanks for the update concerning the client call! 🙂
  
Regarding the struct, I meant we could abstract the validations using a struct that holds the query results and the defined expectations. However, as discussed offline, this can be done in a second iteration, the `spec.TestNRQLExpectedResult` keeps growing and, consequently, we need to add more checks there.

File: internal/spec/definition.go
Comment (sigilioso) [reply internal/spec/definition.go:106]: As discussed, the loop is more readable since the yaml unmarshal customization can be a bit tricky.

File: internal/spec/definition.go
Comment (sigilioso) [inline internal/spec/definition.go:127]: Nit: we don't need the index to perform a single `exepctedResult` validation, it could be included in the error message here by wrapping the error.

File: internal/spec/definition.go
Comment (sigilioso) [inline internal/spec/definition.go:115]: Nit: the validation function could be methods. Example:

```go
func(n TestNRQL) validate() error {
// (...)
}
```

**Review by sigilioso — APPROVED** [review]
Great job @dbudziwojskiNR!

A couple of nits aside (take them only if you think It makes sense), this looks good to me!

**Pull request merged**
